### DISCUSSION TITLE
Calling covariance(m) with a one-column column matrix

### DISCUSSION THREAD

**Discussion by PhilippFeO** [davisking/dlib#2965] [General]
Not sure, whether this belongs here but since it is more a question about the behavior than a bug, I chose Discussions rather than Issues.

I wonder if the following behavior of `covariance(m)` is intended:
```c++
int ROWS = 1, COLS = 5;
matrix<dlib::matrix<double, 0, 1>, 0, 1> m;
matrix<double, 5, 1> col;
col = 0, 1, 2, 3, 4;
m.set_size(ROWS);
for (int i = 0; i < ROWS; ++i) {
    m(i).set_size(COLS);
    m(i) = col;
}
matrix<double> covm = dlib::covariance(m);
std::cout << "covm:\n" << covm << std::endl;
```
yields:
```txt
covm:
-nan -nan -nan -nan -nan                                                                                                                                                                    
-nan -nan -nan -nan -nan                                                                                                                                                                            
-nan -nan -nan -nan -nan                                                                                                                                                                            
-nan -nan -nan -nan -nan                                                                                                                                                                            
-nan -nan -nan -nan -nan
```

The covariance itself is not defined for one random variable/measurement, so calculating it with a one-column column matrix doesn't make sense in the first place but wouldn't be a failed assertion better than a `-nan` output? Especially when there are already consistency checks in https://github.com/davisking/dlib/blob/51c7a3597935ed2cb969521a0ada0b7209f78ec2/dlib/matrix/matrix_utilities.h#L1885

**Comment by davisking** [comment]
Yeah there are asserts there that check this.  They are only on in debug mode or if you explicitly turn them on though.  Turning on all the asserts makes things slow though so they are off by default in release mode.  But yeah it's a good idea to run your CI tests with all the asserts turned on. 

See http://dlib.net/metaprogramming.html#DLIB_ASSERT

**Reply by PhilippFeO** [reply]
Thank you, wasn't aware of that. I am sure this is documented, sorry!

**Reply by davisking** [reply]
No worries 😁
